### PR TITLE
Add support for back porting PRs from GitHub or the Private Azure Repos

### PR DESCRIPTION
**Pull request by TravisEz13**
# PR Summary

Add support for back porting PRs from GitHub or the Private Azure Repos

This pull request enhances the `releaseTools` module by adding functionality to handle different remote repository types (GitHub and Azure Repos) and improving the `Invoke-PRBackport` function to support these changes. The most important changes include the addition of a new function to parse upstream information, updates to handle different remote types, and new test cases to validate these changes.

Enhancements to handle different remote repository types:

* [`tools/releaseTools.psm1`](diffhunk://#diff-ba04f14b3aa7dbb2ebc4dc9158fcd7651396acbd2a382b604ed5525d4f4ff89bR726-R793): Added a new `RemoteType` enum and `Get-UpstreamInfo` function to parse and handle upstream remote information for both GitHub and Azure Repos.
* [`tools/releaseTools.psm1`](diffhunk://#diff-ba04f14b3aa7dbb2ebc4dc9158fcd7651396acbd2a382b604ed5525d4f4ff89bL751-R822): Updated the `Invoke-PRBackport` function to utilize the new `Get-UpstreamInfo` function and handle different remote types when creating and pushing branches. [[1]](diffhunk://#diff-ba04f14b3aa7dbb2ebc4dc9158fcd7651396acbd2a382b604ed5525d4f4ff89bL751-R822) [[2]](diffhunk://#diff-ba04f14b3aa7dbb2ebc4dc9158fcd7651396acbd2a382b604ed5525d4f4ff89bL795-R873) [[3]](diffhunk://#diff-ba04f14b3aa7dbb2ebc4dc9158fcd7651396acbd2a382b604ed5525d4f4ff89bL815-R900) [[4]](diffhunk://#diff-ba04f14b3aa7dbb2ebc4dc9158fcd7651396acbd2a382b604ed5525d4f4ff89bL827-R974)
* [`tools/releaseTools.psm1`](diffhunk://#diff-ba04f14b3aa7dbb2ebc4dc9158fcd7651396acbd2a382b604ed5525d4f4ff89bL851-R998): Exported the `Get-UpstreamInfo` function to make it available for other modules.

New test cases:

* [`test/releaseTools/releaseTools.tests.ps1`](diffhunk://#diff-0858c6b60bffa01d40018b911bb7d96e5528a26daefd8ac572bb32908bd5d6f8R1-R78): Added new test cases to validate the `Get-UpstreamInfo` function for different remote configurations, ensuring it correctly parses remote information for GitHub and Azure Repos.
## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: tools/releaseTools.psm1 [hunk 1]
```diff
@@ -723,6 +723,72 @@ function Get-PRBackportReport {
...
+        [Parameter(Mandatory=$true)]
+        [string[]]$Upstream,
+
+        [Parameter(Mandatory=$true)]
+        [string]$UpstreamRemote
+    )
+
+    $upstreamName = '(powershell(core)?)(/_git)?/(powershell)'
+    $pattern = "^$UpstreamRemote\s*(.*)\:(.*/([-\w.]+)/)?$upstreamName(\.git)?.*fetch"
+    #(.*/([-\w.]+))?(PowerShell(core)?)(/_git)?/(PowerShell)(\.git)?.*fetch
```
Comment (TravisEz13) [inline tools/releaseTools.psm1:742]:
```suggestion
```

File: tools/releaseTools.psm1 [hunk 2]
```diff
@@ -723,6 +723,71 @@ function Get-PRBackportReport {
...
+        Write-Verbose 'parsing Azure repo remote' -Verbose
+        # Azure Repo remote
+        $project = $matches[4]
+        $repo = $matches[7]
+        $upstreamHost = $matches[1]
+
+        if ($upstreamHost -eq 'https') {
+            $upstreamHost = $org
+        }
+        if ($org -match '([^\..]*)\.') {
```
Comment (TravisEz13) [inline tools/releaseTools.psm1:763]:
```suggestion
        # matches everything but `.` ending in a `.`
        # in other word, matching the first part of a hostname.
        if ($org -match '([^\..]*)\.') {
```

**Review by JamesWTruher — APPROVED** [review]
this seems fine. I would probably not provide `$Upstream` as a string that needs parsing, but the component parts you need (like remote name and url as separate elements)

File: tools/releaseTools.psm1 [hunk 3]
```diff
@@ -723,6 +723,74 @@ function Get-PRBackportReport {
...
+        [string[]]$Upstream,
+
+        [Parameter(Mandatory=$true)]
+        [string]$UpstreamRemote
+    )
+
+    $upstreamName = '(powershell(core)?)(/_git)?/(powershell)'
+    $pattern = "^$UpstreamRemote\s*(.*)\:(.*/([-\w.]+)/)?$upstreamName(\.git)?.*fetch"
+    Write-Verbose -Verbose "searching for an upstream with regex: '$pattern'"
+    $upstream = $upstream | Where-Object { $_ -match $pattern }
```
Comment (daxian-dbw) [inline tools/releaseTools.psm1:743]:
```suggestion
    $Upstream = $Upstream | Where-Object { $_ -match $pattern }
```

File: tools/releaseTools.psm1 [hunk 4]
```diff
@@ -723,6 +723,74 @@ function Get-PRBackportReport {
...
+
+    Write-Verbose -Verbose "found $upstream"
+
+    if (!$upstream) {
```
Comment (daxian-dbw) [inline tools/releaseTools.psm1:747]:
```suggestion
    Write-Verbose -Verbose "found $Upstream"

    if (!$Upstream) {
```

File: tools/releaseTools.psm1 [hunk 5]
```diff
@@ -723,6 +723,74 @@ function Get-PRBackportReport {
...
+        throw "Please create an upstream remote that points to $upstreamName"
+    }
+
+    $matches | Format-Table | Out-String -Stream -Width 9999 | Write-Verbose
```
Comment (daxian-dbw) [inline tools/releaseTools.psm1:751]: `$matches` will only contain the matching results from the last item of `$Upstream` that matches `$pattern`, but by looking at the code, there could be multiple items in `$Upstream` that match `$pattern`.

**Review by adityapatwardhan — APPROVED** [review]

**Pull request merged**